### PR TITLE
fix(windows): Start Keyman on Demand - tasks

### LINKED ISSUE #3562: feat(windows): help user start Keyman if it isn't running
With Keyman 14, keyboard TIPs are always registered and installed; if the user selects a Keyman keyboard when Keyman is not running, the TIP attempts to start Keyman, and if it can't, prompts the user to start Keyman.

Relates to #3512.

### PR DESCRIPTION
**Pull request by mcdurdin**
Fixes #3562.

Second part of the two-part special with #3772. When Keyman Configuration starts, it creates a task that is launched whenever a specific Keyman event with id 256 is added to the Event Log. This task launches keyman.exe to start Keyman (and if Keyman is already running, the newly launched process will exit with no side effects).

This feature is behind a new feature flag `Flag_UseAutoStartTask`, so it can be disabled in the event of compatibility issues.

The task is created or updated on each launch of Keyman Configuration (or Keyman via the 'Keyman' shortcut in the Start Menu), and is intended to be self-maintaining.

The test_starttask project (4 of the 12 files) is just a wrapper around the start TKeymanStartTask class to allow for manual testing. No unit tests for this at this time; I think it would be difficult to unit test this effectively.

### CODE REVIEW

**Review by jahorton — COMMENTED** [review]
What I see seems fine codewise... though I really don't get how the task system works.  Do you want me to review that aspect in detail?  If so, I could really use a reference for the task scheduler system & API so that I can make sense of how this works.

I did a few searches to try finding it independently, but never found the same stuff I see used here.

File: windows/src/desktop/kmshell/util/Keyman.System.KeymanStartTask.pas [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+const
+  CTaskName = 'Start On Demand';
+  CTaskDescription = 'Starts Keyman on demand when a Keyman input method is selected';
+  CTaskAuthor = 'SIL International';
+  CTaskFolderName = 'Keyman';
+  CTaskTrigger =
+    '<QueryList>'+
+      '<Query Id="0" Path="Application">'+
+        '<Select Path="Application">*[System[Provider[@Name=''Keyman''] and EventID=256]]</Select>'+
```
Comment (darcywong00) [inline windows/src/desktop/kmshell/util/Keyman.System.KeymanStartTask.pas:39]: > it creates a task that is launched whenever a specific Keyman event with id 256 is added to the Event Log

Of all the id's, why `256`?

File: windows/src/global/delphi/general/Keyman.System.Settings.pas [hunk 2]
```diff
@@ -409,6 +409,20 @@   TKeymanSettings = class(TObjectList<TKeymanSetting>)
       ValueType: kstInteger
     ),
 
+    // kmshell:Keyman.System.KeymanStartTask
+    (
+      ID: 'engine.compatibility.use_autostart_task';
```
Comment (darcywong00) [inline windows/src/global/delphi/general/Keyman.System.Settings.pas:414]: Is this tied to the Keyman configuration option "Start when Windows starts"?

**Comment by mcdurdin** [comment]
> What I see seems fine codewise... though I really don't get how the task system works. Do you want me to review that aspect in detail? If so, I could really use a reference for the task scheduler system & API so that I can make sense of how this works.

I'm not sure how much detail we need to go into. A basic understanding of it is helpful. I worked from this code sample as a basis: https://docs.microsoft.com/en-us/previous-versions/aa446886(v=vs.85)

File: windows/src/desktop/kmshell/util/Keyman.System.KeymanStartTask.pas (hunk 1 above)
Comment (mcdurdin) [reply windows/src/desktop/kmshell/util/Keyman.System.KeymanStartTask.pas:39]: `0x100` was the event id used in the example. It's just a number, not important.

File: windows/src/global/delphi/general/Keyman.System.Settings.pas (hunk 2 above)
Comment (mcdurdin) [reply windows/src/global/delphi/general/Keyman.System.Settings.pas:414]: No, it's unrelated. This autostart only happens when the user selects a Keyman keyboard using the Windows language selector, and Keyman is not already running.

**Review by darcywong00 — APPROVED** [review]
lgtm

**Comment by jahorton** [comment]
I guess the part I'm missing is "what code calls `CreateTask`?"  It's a new function, but outside of the new files and the new manual-test, I don't see any references to it.

Given that we're talking about the TIP starting Keyman on demand, I expected to see _something_ altered within the `/kmtip/` subdirectory, but maybe my expectation is invalid.

**Comment by jahorton** [comment]
> Fixes #3562.
> 
> Second part of the two-part special with #3772. When Keyman Configuration starts, it creates a task that is launched whenever a specific Keyman event with id 256 is added to the Event Log. This task launches keyman.exe to start Keyman (and if Keyman is already running, the newly launched process will exit with no side effects).
> 
> This feature is behind a new feature flag `Flag_UseAutoStartTask`, so it can be disabled in the event of compatibility issues.
> 
> The task is created or updated on each launch of Keyman Configuration (or Keyman via the 'Keyman' shortcut in the Start Menu), and is intended to be self-maintaining.
> 
> The test_starttask project (4 of the 12 files) is just a wrapper around the start TKeymanStartTask class to allow for manual testing. No unit tests for this at this time; I think it would be difficult to unit test this effectively.

Alternatively, I do see this note... but don't see any obvious matching edits within the `kmconfig` section either.

**Review by jahorton — APPROVED** [review]

**Pull request merged**